src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs

```diff
@@ -0,0 +1,417 @@
+        /// <param name="context">Serializer context</param>
+        /// <returns>Returns the URI for NextPageLink. If a null object is passed for the instance, resorts to the default paging mechanism of using $skip and $top.</returns>
+        public override Uri GenerateNextPageLink(Uri baseUri, int pageSize, Object instance, ODataSerializerContext context)
+        {
+            if (context == null)
+            {
+                return null;
+            }
+
+            if (pageSize <= 0)
+            {
+                return null;
+            }
+
+            Func<object, string> skipTokenGenerator = null;
+            IList<OrderByNode> orderByNodes = null;
+            ExpandedReferenceSelectItem expandedItem = context.CurrentExpandedSelectItem;
+            IEdmModel model = context.Model;
+
+            DefaultQuerySettings settings = context.QueryContext.DefaultQuerySettings;
+            if (settings.EnableSkipToken)
+            {
+                if (expandedItem != null)
+                {
+                    // Handle Delta resource; currently not value based.
+                    if (TypedDelta.IsDeltaOfT(context.ExpandedResource.GetType()))
+                    {
+                        return GetNextPageHelper.GetNextPageLink(baseUri, pageSize);
+                    }
+
+                    if (expandedItem.OrderByOption != null)
+                    {
+                        orderByNodes = OrderByNode.CreateCollection(expandedItem.OrderByOption);
+                    }
+
+                    skipTokenGenerator = (obj) =>
+                    {
+                        return GenerateSkipTokenValue(obj, model, orderByNodes);
+                    };
+
+                    return GetNextPageHelper.GetNextPageLink(baseUri, pageSize, instance, skipTokenGenerator);
+                }
+
+                if (context.QueryOptions != null && context.QueryOptions.OrderBy != null)
+                {
+                    orderByNodes = context.QueryOptions.OrderBy.OrderByNodes;
+                }
+
+                skipTokenGenerator = (obj) =>
+                {
+                    return GenerateSkipTokenValue(obj, model, orderByNodes);
+                };
+            }
+
+            return context.InternalRequest.GetNextPageLink(pageSize, instance, skipTokenGenerator);
+        }
+
+        /// <summary>
+        /// Returns a string that converts an object to a skiptoken value string
+        /// </summary>
+        /// <param name="lastMember"> Object based on which SkipToken value will be generated.</param>
+        /// <param name="model">The edm model.</param>
+        /// <param name="orderByNodes">List of orderByNodes used to generate the skiptoken value.</param>
+        /// <returns>Value for the skiptoken to be used in the next link.</returns>
+        protected virtual string GenerateSkipTokenValue(Object lastMember, IEdmModel model, IList<OrderByNode> orderByNodes)
```

Comment:
>IList<OrderByNode> orderByNodes [](start = 92, length = 31)

I thought we had updated to take QueryOptions?

Reply:
This is not part of the public interface for generating the next link. Rather this is an extensible virtual method on the default implementation class. That is, if someone wants to use a similar enough implementation, they can use override this method. Moreover, QueryOptions will not fly here as for the nested resource there are no query options object. 

---
In reply to: [293021744](https://github.com/OData/WebApi/pull/1724#discussion_r293021744) [](ancestors = 293021744)